ampd/src/starknet/types/array_span.rs

```diff
@@ -0,0 +1,285 @@
+///     FieldElement::from_str(
+///         "0x000000000000000000000000000000000000000000000000000000000000006c",
+///     )
+///     .unwrap(),
+///     FieldElement::from_str(
+///         "0x000000000000000000000000000000000000000000000000000000000000006f",
+///     )
+///     .unwrap(),
+/// ];
+///
+/// let array_span = ArraySpan::try_from(data).unwrap();
+/// assert_eq!(array_span.bytes, vec![104, 101, 108, 108, 111]);
+/// assert_eq!(String::from_utf8(array_span.bytes).unwrap(), "hello");
+/// ```
+///
+/// For more info:
+/// https://docs.starknet.io/documentation/architecture_and_concepts/Smart_Contracts/serialization_of_Cairo_types/#serialization_of_byte_arrays
+#[derive(Debug)]
+pub struct ArraySpan {
+    pub bytes: Vec<u8>,
+}
+
+#[derive(Error, Debug)]
+pub enum ArraySpanError {
+    #[error("Invalid array/span length")]
+    InvalidLength,
+    #[error("Failed to parse felt - {0}")]
+    ParsingFelt(#[from] ValueOutOfRangeError),
+}
+
+impl TryFrom<Vec<FieldElement>> for ArraySpan {
+    type Error = ArraySpanError;
+
+    fn try_from(data: Vec<FieldElement>) -> Result<Self, Self::Error> {
+        // First element is always the array length.
+        // We also have to go from `u32` to usize, because
+        // there's no direct `usize` From impl.
+        let arr_length = u32::try_from(data[0])?;
+
+        // -1 because we have to offset the first element (the length itself)
+        let is_arr_el_count_valid = usize::try_from(arr_length)
+            .map(|count| count == data.len() - 1)
+            .unwrap_or(false);
+
+        if !is_arr_el_count_valid {
+            return Err(ArraySpanError::InvalidLength);
+        }
+
+        let bytes: Result<Vec<u8>, ArraySpanError> = match data.get(1..) {
+            Some(b) => b,
+            None => return Err(ArraySpanError::InvalidLength),
+        }
+        .to_vec()
+        .into_iter()
+        .map(|e| {
+            let word_count: u8 = match e.try_into() {
+                Ok(wc) => wc,
+                Err(err) => return Err(ArraySpanError::ParsingFelt(err)),
+            };
+
+            Ok(word_count)
+        })
+        .collect();
+
+        Ok(ArraySpan { bytes: bytes? })
+    }
+}
+
+#[cfg(test)]
+mod array_span_tests {
+    use std::str::FromStr;
+
+    use starknet_core::types::FieldElement;
+
+    use crate::starknet::types::array_span::ArraySpan;
+
+    #[test]
+    fn try_from_valid_zeros() {
+        // the string "hello", but FieldElement is bigger than u8::max
+        let data = vec![FieldElement::from_str(
+            "0x0000000000000000000000000000000000000000000000000000000000000000",
+        )
+        .unwrap()];
+
+        let array_span = ArraySpan::try_from(data).unwrap();
+        assert_eq!(array_span.bytes, Vec::<u8>::new());
+    }
+
+    #[test]
+    fn try_from_failed_to_parse_element_to_u8() {
+        // the string "hello", but FieldElement is bigger than u8::max
+        let data = vec![
+            FieldElement::from_str(
```

Comment:
It looks like we could avoid a lot of repetition here and in the other methods by doing something like:

```
let result: Result<Vec<FieldElement>, from_str_error_type_here> = vec![
"0x000....",
"0x000....",
...
].into_iter().map(FieldElement::from_str).collect();

result.unwrap();
```

And then even further by creating a method that does this and just reusing that method in all test cases.

Reply:
Nice suggestion. Refactored it.